In [2]:
import pandas as pd
import numpy as np
from sklearn.feature_selection import SelectPercentile, f_classif
from sklearn.naive_bayes import GaussianNB


In [4]:
import nltk
from sklearn.feature_extraction.text import TfidfVectorizer

In [5]:
df =pd.read_csv("inshorts data preprocessed.csv")

In [6]:
df.head()

Subject                                     Headline_lemma  \
0       Politics                 stopped entering studio time arnab   
1  Entertainment                new trailer justice league released   
2  Entertainment        touch right shilpa shinde sexual harassment   
3       Politics  antiromeo squad must trouble consenting youth ...   
4       Politics            romeo juliet welcome delhi aap minister   

                                         Short_lemma  
0  tv news anchor arnab goswami said told could p...  
1  new trailer upcoming superhero film justice le...  
2  television actress shilpa shinde opening claim...  
3  uttar pradesh chief minister yogi adityanath v...  
4  apparent jibe ups antiromeo squad delhi touris...

In [7]:
df.shape

(25196, 3)

In [25]:
df2 = df.loc[14000:16000,:]

In [26]:
df2.head()

Subject                                     Headline_lemma  \
14000  Politics                 uk ban fishing million sq km ocean   
14001  Politics                 iimc get upgraded university naidu   
14002  Politics                   west bengal give acre land volvo   
14003  Politics  vijaya lakshmi st female un general assembly prez   
14004  Politics              tax board honour lakh honest taxpayer   

                                             Short_lemma  
14000  uk government thursday said ban commercial fis...  
14001  process upgrade indian institute mass communic...  
14002  west bengal chief minister mamata banerjee thu...  
14003  indian diplomat politician vijaya lakshmi pand...  
14004  central board direct tax honour lakh honest co...

In [39]:
vectorizer = TfidfVectorizer()

In [30]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OrdinalEncoder

In [31]:
enc =OrdinalEncoder()

In [44]:
X1 =df2[['Subject']]
enc.fit(X1)

OrdinalEncoder()

In [46]:
df2['Subejct_encoded']=enc.transform(df2[['Subject']])

C:\Users\admin\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [47]:
df2.head()

Subject                                     Headline_lemma  \
14000  Politics                 uk ban fishing million sq km ocean   
14001  Politics                 iimc get upgraded university naidu   
14002  Politics                   west bengal give acre land volvo   
14003  Politics  vijaya lakshmi st female un general assembly prez   
14004  Politics              tax board honour lakh honest taxpayer   

                                             Short_lemma  Subejct_encoded  
14000  uk government thursday said ban commercial fis...              2.0  
14001  process upgrade indian institute mass communic...              2.0  
14002  west bengal chief minister mamata banerjee thu...              2.0  
14003  indian diplomat politician vijaya lakshmi pand...              2.0  
14004  central board direct tax honour lakh honest co...              2.0

In [48]:
X2 = df2['Short_lemma']

In [49]:
y2=df2['Subejct_encoded']

In [52]:
features_train2, features_test2, labels_train2, labels_test2 = train_test_split(X2, y2, test_size=0.33, random_state=42)

In [53]:
features_train2 = vectorizer.fit_transform(features_train2)
features_test2 = vectorizer.transform(features_test2)

In [54]:
model = GaussianNB()
model.fit(features_train2.toarray(), labels_train2)

GaussianNB()

In [56]:
score_train = model.score(features_train2.toarray(), labels_train2)

In [57]:
score_train

1.0

In [59]:
score_test = model.score(features_test2.toarray(), labels_test2)

In [60]:
score_test

0.7760968229954615

In [61]:
y_pred = model.predict(features_test2.toarray())

In [65]:
from sklearn.metrics import confusion_matrix,classification_report

In [64]:
confusion_matrix(labels_test2,y_pred)

array([[ 23,   4,  11,   0,   3],
       [  1,  98,  31,   2,  12],
       [  1,   3, 297,   4,  11],
       [  2,   5,  21,  25,   4],
       [  0,   9,  22,   2,  70]], dtype=int64)

In [66]:
classification_report(labels_test2,y_pred)

'              precision    recall  f1-score   support\n\n         0.0       0.85      0.56      0.68        41\n         1.0       0.82      0.68      0.75       144\n         2.0       0.78      0.94      0.85       316\n         3.0       0.76      0.44      0.56        57\n         4.0       0.70      0.68      0.69       103\n\n    accuracy                           0.78       661\n   macro avg       0.78      0.66      0.70       661\nweighted avg       0.78      0.78      0.77       661\n'

In [68]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print('\nAccuracy: {:.2f}\n'.format(accuracy_score(labels_test2, y_pred)))

print('Micro Precision: {:.2f}'.format(precision_score(labels_test2, y_pred, average='micro')))
print('Micro Recall: {:.2f}'.format(recall_score(labels_test2, y_pred, average='micro')))
print('Micro F1-score: {:.2f}\n'.format(f1_score(labels_test2, y_pred, average='micro')))

print('Macro Precision: {:.2f}'.format(precision_score(labels_test2, y_pred, average='macro')))
print('Macro Recall: {:.2f}'.format(recall_score(labels_test2, y_pred, average='macro')))
print('Macro F1-score: {:.2f}\n'.format(f1_score(labels_test2, y_pred, average='macro')))

print('Weighted Precision: {:.2f}'.format(precision_score(labels_test2, y_pred, average='weighted')))
print('Weighted Recall: {:.2f}'.format(recall_score(labels_test2, y_pred, average='weighted')))
print('Weighted F1-score: {:.2f}'.format(f1_score(labels_test2, y_pred, average='weighted')))

from sklearn.metrics import classification_report
print('\nClassification Report\n')
print(classification_report(labels_test2, y_pred, target_names=['Class 1', 'Class 2', 'Class 3','Class 4','Class 5']))



Accuracy: 0.78

Micro Precision: 0.78
Micro Recall: 0.78
Micro F1-score: 0.78

Macro Precision: 0.78
Macro Recall: 0.66
Macro F1-score: 0.70

Weighted Precision: 0.78
Weighted Recall: 0.78
Weighted F1-score: 0.77

Classification Report

              precision    recall  f1-score   support

     Class 1       0.85      0.56      0.68        41
     Class 2       0.82      0.68      0.75       144
     Class 3       0.78      0.94      0.85       316
     Class 4       0.76      0.44      0.56        57
     Class 5       0.70      0.68      0.69       103

    accuracy                           0.78       661
   macro avg       0.78      0.66      0.70       661
weighted avg       0.78      0.78      0.77       661



In [70]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(max_features='auto',random_state=1, n_jobs=-1)


In [71]:
from sklearn.model_selection import GridSearchCV


In [72]:
param_grid = { "criterion" : ["gini", "entropy"], "min_samples_leaf" : [1, 5, 10], "min_samples_split" : [2, 4, 10,12,16], "n_estimators": [50, 100,400,700,1000]}
    

In [74]:
gs = GridSearchCV(estimator=rf, param_grid=param_grid, scoring='accuracy', cv=3, n_jobs=-1)
gs = gs.fit(features_train2.toarray(), labels_train2)


In [75]:
print(gs.best_estimator_) 
print(gs.best_score_)

RandomForestClassifier(n_estimators=1000, n_jobs=-1, random_state=1)
0.916418374615022


In [80]:
from sklearn.model_selection import cross_val_score

In [82]:
rf = RandomForestClassifier(n_estimators=1000, n_jobs=-1, random_state=1)
rf.fit(features_train2.toarray(), labels_train2)
y_predictions_rf = rf.predict(features_test2.toarray())
acc = accuracy_score(labels_test2, y_predictions_rf)
print(acc)

0.9046898638426626


In [83]:
confusion_matrix(labels_test2,y_predictions_rf)

array([[ 39,   0,   2,   0,   0],
       [  0, 122,  18,   1,   3],
       [  0,   1, 310,   2,   3],
       [  0,   2,   9,  46,   0],
       [  0,   2,  20,   0,  81]], dtype=int64)

In [85]:
print('\nAccuracy: {:.2f}\n'.format(accuracy_score(labels_test2, y_predictions_rf)))

print('Micro Precision: {:.2f}'.format(precision_score(labels_test2, y_pred, average='micro')))
print('Micro Recall: {:.2f}'.format(recall_score(labels_test2, y_predictions_rf, average='micro')))
print('Micro F1-score: {:.2f}\n'.format(f1_score(labels_test2, y_predictions_rf, average='micro')))

print('Macro Precision: {:.2f}'.format(precision_score(labels_test2, y_predictions_rf, average='macro')))
print('Macro Recall: {:.2f}'.format(recall_score(labels_test2, y_predictions_rf, average='macro')))
print('Macro F1-score: {:.2f}\n'.format(f1_score(labels_test2, y_predictions_rf, average='macro')))

print('Weighted Precision: {:.2f}'.format(precision_score(labels_test2, y_predictions_rf, average='weighted')))
print('Weighted Recall: {:.2f}'.format(recall_score(labels_test2, y_predictions_rf, average='weighted')))
print('Weighted F1-score: {:.2f}'.format(f1_score(labels_test2, y_predictions_rf, average='weighted')))

print('\nClassification Report\n')
print(classification_report(labels_test2, y_predictions_rf, target_names=['Class 1', 'Class 2', 'Class 3','Class 4','Class 5']))



Accuracy: 0.90

Micro Precision: 0.78
Micro Recall: 0.90
Micro F1-score: 0.90

Macro Precision: 0.94
Macro Recall: 0.87
Macro F1-score: 0.90

Weighted Precision: 0.91
Weighted Recall: 0.90
Weighted F1-score: 0.90

Classification Report

              precision    recall  f1-score   support

     Class 1       1.00      0.95      0.97        41
     Class 2       0.96      0.85      0.90       144
     Class 3       0.86      0.98      0.92       316
     Class 4       0.94      0.81      0.87        57
     Class 5       0.93      0.79      0.85       103

    accuracy                           0.90       661
   macro avg       0.94      0.87      0.90       661
weighted avg       0.91      0.90      0.90       661



In [86]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(p=2,weights='distance',n_neighbors=50)
knn.fit(features_train2.toarray(), labels_train2)

# Predict using model:

y_predict_knn=knn.predict(features_test2.toarray())



In [87]:
print('\nAccuracy: {:.2f}\n'.format(accuracy_score(labels_test2, y_predict_knn)))

print('Micro Precision: {:.2f}'.format(precision_score(labels_test2, y_predict_knn, average='micro')))
print('Micro Recall: {:.2f}'.format(recall_score(labels_test2, y_predict_knn, average='micro')))
print('Micro F1-score: {:.2f}\n'.format(f1_score(labels_test2, y_predict_knn, average='micro')))

print('Macro Precision: {:.2f}'.format(precision_score(labels_test2, y_predict_knn, average='macro')))
print('Macro Recall: {:.2f}'.format(recall_score(labels_test2, y_predict_knn, average='macro')))
print('Macro F1-score: {:.2f}\n'.format(f1_score(labels_test2, y_predict_knn, average='macro')))

print('Weighted Precision: {:.2f}'.format(precision_score(labels_test2, y_predict_knn, average='weighted')))
print('Weighted Recall: {:.2f}'.format(recall_score(labels_test2, y_predict_knn, average='weighted')))
print('Weighted F1-score: {:.2f}'.format(f1_score(labels_test2, y_predict_knn, average='weighted')))

print('\nClassification Report\n')
print(classification_report(labels_test2, y_predict_knn, target_names=['Class 1', 'Class 2', 'Class 3','Class 4','Class 5']))



Accuracy: 0.85

Micro Precision: 0.85
Micro Recall: 0.85
Micro F1-score: 0.85

Macro Precision: 0.89
Macro Recall: 0.76
Macro F1-score: 0.81

Weighted Precision: 0.86
Weighted Recall: 0.85
Weighted F1-score: 0.84

Classification Report

              precision    recall  f1-score   support

     Class 1       0.94      0.78      0.85        41
     Class 2       0.90      0.78      0.84       144
     Class 3       0.81      0.98      0.89       316
     Class 4       0.93      0.49      0.64        57
     Class 5       0.87      0.76      0.81       103

    accuracy                           0.85       661
   macro avg       0.89      0.76      0.81       661
weighted avg       0.86      0.85      0.84       661



In [88]:
print('\nAccuracy of Naive Bayes Model: {:.2f}\n'.format(accuracy_score(labels_test2, y_pred)))
print('\nAccuracy of KNN Model: {:.2f}\n'.format(accuracy_score(labels_test2, y_predict_knn)))
print('\nAccuracy of Random Forest Model: {:.2f}\n'.format(accuracy_score(labels_test2, y_predictions_rf)))



Accuracy of Naive Bayes Model: 0.78


Accuracy of KNN Model: 0.85


Accuracy of Random Forest Model: 0.90



In [89]:
print('Weighted F1-score of Naive Bayes Model: {:.2f}'.format(f1_score(labels_test2, y_pred, average='weighted')))
print('Weighted F1-score of KNN Model: {:.2f}'.format(f1_score(labels_test2, y_predict_knn, average='weighted')))
print('Weighted F1-score of Random Forest: {:.2f}'.format(f1_score(labels_test2, y_predictions_rf, average='weighted')))


Weighted F1-score of Naive Bayes Model: 0.77
Weighted F1-score of KNN Model: 0.84
Weighted F1-score of Random Forest: 0.90


As we can see from our analysis that random forest model performs the best amongst the three chosen models. But one noticeable aspect here is the performance of a simple KNN model. With accuracy of 85% it is not much behind.